### ISSUE TITLE
LSP not responding when editor quits and sends shutdown request

### ISSUE THREAD

**Issue by cngarrison** [denoland/deno#20700]
When quitting my EDITOR ([BBEdit](https://www.barebones.com/products/bbedit/)), it hangs waiting for `deno lsp`. 

I have reported the issue to BareBones. They advised that it's possibly due to LSP server not responding correctly to the application's request to shut itself down. 

- https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#shutdown

Does `deno lsp` respond to the `shutdown` request?

Is there a config setting for handling the `shutdown` request?

**Comment by nayeemrmn** [comment]
The shutdown handler we have just tries to acquire the global mutex it has (ensuring in-flight operations finish) and then responds with `Ok(())`. If it's really hanging, maybe it's because a previous request going either way is on-going. I don't think this is been a problem with other editors though. Unfortunately I can't investigate very far since BBEdit's source isn't public.

@cngarrison Does this happen with a bare minimum reproduction? E.g. open an empty ts file in a fresh directory, plus whichever method you're using to enable the Deno LSP, and try to exit.

**Comment by cngarrison** [comment]
@nayeemrmn Thanks for that feedback.

I launched BBEdit with `deno` LSP server enabled and no documents open. I then quit BBEdit and it quit cleanly.

I launched BBEdit, opened a new project directory (file browser), created an (empty) index.ts file, then quit BBEdit. The hang returned. 

The team at BareBones is very good, and I have flagged this issue with them. Please advise if there is any further testing I can help with.

**Comment by nayeemrmn** [comment]
> When quitting my EDITOR ([BBEdit](https://www.barebones.com/products/bbedit/)), it hangs waiting for `deno lsp`.
> 
> I have reported the issue to BareBones. They advised that it's possibly due to LSP server not responding correctly to the application's request to shut itself down.

@sigmaSd Do you know of any quirks that might relate to this?

**Comment by siegel** [comment]
For background (and in case it helps focus the investigation):

When stopping a language server (at application quit or for certain configuration changes), BBEdit sends a `shutdown` request to the server and waits for a response.

After receiving a response to `shutdown`, BBEdit then sends the server an `exit` notification.

Then, BBEdit waits for the server process to exit. This is done because some language servers actually do nothing at `shutdown` time, and instead do all of their cleanup (index database commits, and the like) when they receive an `exit`, and in some cases this can take a number of seconds. (An example of this is found in [clangd](https://github.com/clangd/clangd/issues/855 .)

In the case of `deno`, it responds normally (and rapidly) to `shutdown`, but after receiving the `exit` notification, the `deno` process never terminates.

If `deno` has any work to do when shutting the server down (such as committing an index database, or other housekeeping), the `shutdown` handler is the time to do this; and the `exit` notification handler simply needs to terminate the server process.

Hope this helps. :-)

**Comment by sigmaSd** [comment]
Its never been a problem on nvim/helix, so I don't have other informations regarding this.

**Comment by nayeemrmn** [comment]
@siegel Is there a change BBEdit isn't closing the LSP process's stdin?

**Comment by siegel** [comment]
> @siegel Is there a chance BBEdit isn't closing the LSP process's stdin?

Probably. :-) BBEdit does not explicitly close the pipes until after the server process has exited as requested.

Is there something in the LSP specification that states that the client is expected to close `stdin` after sending the `exit` notification?

**Comment by nayeemrmn** [comment]
> Is there something in the LSP specification that states that the client is expected to close `stdin` after sending the `exit` notification?

Nope. Based on the wording at https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#exit we should exit without needing that. But given that this issue occurs specifically between Deno and BBEdit, maybe both actions should be taken

@ebkalderon Could `Server::serve()` be changed to break its loop on exit notifications without waiting for stdin to close, so our program can finish up as normal? Or is there something we can do currently?

**Comment by siegel** [comment]
> Please feel free to open a bug report on the `tower-lsp` repo to investigate further. Thanks for bringing this to my attention, folks!

Done! [Your #399](https://github.com/ebkalderon/tower-lsp/issues/399) is it. And I appreciate you taking an interest in looking into it. :-)

**Comment by jeffs** [comment]
This seems to be happening for me with 2.7.5 and 2.7.6. The LSP server hangs frequently in general; but the most visible upshot is that when I quit Helix, there's a 30s hang while Helix waits in vain for deno-lsp to shut down.
